### PR TITLE
Add `QualityScorer` Task

### LINKED ISSUE #390: Adapt `QualityScorerTask` to new `Task` interface

### PR DESCRIPTION
**Pull request by ignacioct**
Closes #390 

This PR adds the `QualityScorer` task described in the [DEITA paper]. Thanks a lot to Gabri for guiding me and for the PR on the `ComplexityScorer`, both are really similar.

### CODE REVIEW

File: src/distilabel/steps/task/quality_scorer.py [hunk 1]
```diff
@@ -64,7 +66,7 @@ def load(self) -> None:
     @property
     def inputs(self) -> List[str]:
         """The input for the task is the `instruction`."""
```
Comment (alvarobartt) [inline src/distilabel/steps/task/quality_scorer.py:68]:
```suggestion
        """The input for the task are `instruction` and `responses`."""
```

File: src/distilabel/steps/task/quality_scorer.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    QualityScorerTask is a pre-defined task that defines the `instruction` as the input
+    and `score` as the output. This task is used to rate the quality of instructions and responses.
+    It is inspired by the Evol Quality Scorer in the Deita framework: : *Deita is an open-sourced project
+    designed to facilitate Automatic Data Selection for instruction tuning in Large Language Models (LLMs).*
+    The task follows the same scheme as the Evol Complexity Scorer, but the instruction-response pairs are scored in
+    terms of quality, obtining a quality score *q* for each instruction.
+
+    Input columns:
+        instruction (`List[str]`): The instruction to be scored.
+        responses (`List[str]`): The responses to that instruction to be scored. Each response forms a pair with the instruction.
```
Comment (gabrielmbmb) [inline src/distilabel/steps/task/quality_scorer.py:51]:
```suggestion
        instruction (`str`): The instruction that was used to generate the `responses`.
        responses (`List[str]`): The responses to be scored. Each response forms a pair with the instruction.
```

File: src/distilabel/steps/task/quality_scorer.py [hunk 3]
```diff
@@ -75,21 +77,22 @@ def format_input(self, input: str) -> ChatType:  # type: ignore
     @property
     def outputs(self):
         """The output for the task is the `generation` and the `model_name`."""
-        return ["quality_score"]
+        return ["quality_scores"]
```
Comment (alvarobartt) [inline src/distilabel/steps/task/quality_scorer.py:80]: Maybe we can just leave it as `scores`? Note that the user will always have the possibility to rename it to `quality_scores` or any other name if `scores` is confusing, while keeping `scores` for the most common scenarios 👍🏻 

```suggestion
        return ["scores"]
```

File: src/distilabel/steps/task/quality_scorer.py [hunk 4]
```diff
@@ -75,21 +77,22 @@ def format_input(self, input: str) -> ChatType:  # type: ignore
...
 
     def format_output(
         self, output: Union[str, None], input: Dict[str, Any]
     ) -> Dict[str, Any]:
-        """The output is formatted as a list with the rating of each instruction.
+        """The output is formatted as a list with the rating of each instruction-response tuple.
```
Comment (alvarobartt) [inline src/distilabel/steps/task/quality_scorer.py:85]:
```suggestion
        """The output is formatted as a list with the score of each instruction-response pair.
```

File: src/distilabel/steps/task/quality_scorer.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    @property
+    def inputs(self) -> List[str]:
+        """The input for the task is the `instruction`."""
+        return ["instruction", "responses"]
+
+    def format_input(self, input: str) -> ChatType:  # type: ignore
+        """The input is formatted as a `ChatType` assuming that the instruction
+        is the first interaction from the user within a conversation. And the
+        `system_prompt` is added as the first message if it exists."""
```
Comment (gabrielmbmb) [inline src/distilabel/steps/task/quality_scorer.py:74]:
```suggestion
        """The input is formatted as a `ChatType` assuming that the instruction
        is the first interaction from the user within a conversation."""
```

File: src/distilabel/steps/task/quality_scorer.py [hunk 6]
```diff
@@ -75,21 +77,22 @@ def format_input(self, input: str) -> ChatType:  # type: ignore
...
+        return ["quality_scores"]
 
     def format_output(
         self, output: Union[str, None], input: Dict[str, Any]
     ) -> Dict[str, Any]:
-        """The output is formatted as a list with the rating of each instruction.
+        """The output is formatted as a list with the rating of each instruction-response tuple.
 
         Args:
             output (str): the raw output of the LLM.
```
Comment (alvarobartt) [inline src/distilabel/steps/task/quality_scorer.py:88]: We decided not to include the type-hints within the docstrings, as those will be automatically inferred from the function/method signature, so feel free to remove those from the docstrings

```suggestion
            output: the raw output of the LLM.
```

File: src/distilabel/steps/task/quality_scorer.py [hunk 7]
```diff
@@ -75,21 +77,22 @@ def format_input(self, input: str) -> ChatType:  # type: ignore
...
 
         Returns:
-            Dict[str, List[str]]: A dict with containing the ratings of each instruction."""
+            Dict[str, List[str]]: A dict with containing the ratings of each instruction-response.
```
Comment (alvarobartt) [inline src/distilabel/steps/task/quality_scorer.py:91]: Same here!

```suggestion
            A dict with containing the scores for each instruction-response pair.
```

File: tests/unit/steps/task/test_quality_scorer.py [hunk 8]
```diff
@@ -42,15 +45,15 @@ def test_format_input(self) -> None:
         [
             (
                 "[1] Score: 1\n[2] Score: 2\n[3] Score: 3\n",
-                {"quality_score": [1.0, 2.0, 3.0]},
+                {"quality_scores": [1.0, 2.0, 3.0]},
```
Comment (alvarobartt) [inline tests/unit/steps/task/test_quality_scorer.py:48]: If you happen to apply the change above to rename `quality_scores` to `scores`, then it should be updated within this file too 👍🏻

File: src/distilabel/steps/task/quality_scorer.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class QualityScorer(Task):
+    """
+    QualityScorerTask is a pre-defined task that defines the `instruction` as the input
+    and `score` as the output. This task is used to rate the quality of instructions and responses.
+    It is inspired by the Evol Quality Scorer in the Deita framework: : *Deita is an open-sourced project
+    designed to facilitate Automatic Data Selection for instruction tuning in Large Language Models (LLMs).*
+    The task follows the same scheme as the Evol Complexity Scorer, but the instruction-response pairs are scored in
+    terms of quality, obtining a quality score *q* for each instruction.
```
Comment (gabrielmbmb) [inline src/distilabel/steps/task/quality_scorer.py:47]:
```suggestion
    QualityScorerTask is a pre-defined task that defines the `instruction` as the input
    and `score` as the output. This task is used to rate the quality of instructions and responses.
    It's an implementation of the quality score task from the paper 'What Makes Good Data
    for Alignment? A Comprehensive Study of Automatic Data Selection in Instruction Tuning'.
    The task follows the same scheme as the Complexity Scorer, but the instruction-response pairs
    are scored in terms of quality, obtaining a quality score for each instruction.
```

**Review by gabrielmbmb — CHANGES_REQUESTED** [review]
LGTM! Just a few changes in the docstrings and should be ready.

**Review by alvarobartt — APPROVED** [review]

**Comment by alvarobartt** [comment]
LGTM @ignacioct great job! BTW could you please update the PR description to be similar to https://github.com/argilla-io/distilabel/pull/407?

Also would be cool if you could share a simple but working example (feel free to reuse the code from the linked PR above!) 👍🏻 Then, feel free to merge it already!

**Pull request merged**